addons/mail/controllers/mail.py

```diff
@@ -197,3 +201,53 @@ def mail_action_unfollow(self, model, res_id, pid, token, **kwargs):
             'model_name': request.env['ir.model'].sudo()._get(model).display_name,
             'access_url': record._notify_get_action_link('view', model=model, res_id=res_id) if display_link else False,
         })
+
+    @http.route("/mail/message/redirect", methods=["POST"], type="json", auth="public")
+    @add_guest_to_context
+    def mail_message_redirect(self, message_id):
+        message = request.env['mail.message'].search([('id', '=', message_id)])
+        if not message:
+            return {'error': 'Unauthorized'}
+        # sudo: public user can access some relational fields of mail.message
+        if message.sudo()._filter_empty():
+            return {'error': 'NotFound'}
+        store = Store()
+        if message.model == 'discuss.channel':
+            store.add(request.env['discuss.channel'].browse(message.res_id))
+        store.add('Thread', {'id': message.res_id, 'model': message.model, 'highlightMessageId': message_id})
+        return {'threadData': store.get_result()}
+
+    @http.route('/mail/<string:res_model>/<int:res_id>/message/redirect/<int:message_id>', type='http', auth='public')
+    @add_guest_to_context
+    def mail_thread_message_redirect(self, res_model, res_id, message_id, **kwargs):
```

Comment:
Well It could misguide that the route fetches data for message "1". But, I think it makes the route unnecessarily longer. So, I would remove the `redirect` part.